Yiz/modify list parents

sleepyzzzzzz opened the pull request:
## Changes made in this PR
ListParentsByID:
change to returning partially valid parents and no error when parent ids given are partially valid and existing

ListAllParents:
1. change the parents list order by ascending id
2. modify next page token value in grpc
3. modify tests in grpc + store layers

## Corresponding noted Ruby branch

< Please go to the paramount-api repo, create a separate pull request, add comment on the functionality that's re-implemented in this the go codebase
For example: https://github.com/ParamountDataManagement/paramount-api/compare/lhan/user-notes?expand=1 >

## TODO - changes that needs to be made
< leave the todos here to help the reviewers know what's the remaining pieces that haven't been implemented >
                                                                                        
## Implementation Checklist
Please check following check list to ensure the tasks are considered and completed in this branch.

#### API Design
- [ ] Defined gRPC API in proto file
- [ ] regenerated go code using `make proto_generate` command

#### API implementation
- [ ] Implemented API in corresponding gRPC server
- [ ] Added tests in gRPC layer

#### Domain Service and DB query Implementation
- [ ] Added new query in dbstore layer
- [ ] Added tests in db store layer
                                                                                       
- [ ] Implemented functionality in Domain Service Layer
- [ ] Added test in Domain Service Layer

- [ ] Created a new branch in paramount-api, with comments there indicating the functionality in Ruby that's re-implemented in Golang

#### Postgres Layer - identify and potentially add db migration to create index for fast querying.
- [ ] Verified and potentially created index for the newly added query.

lhan-the-real commented on internal/domain/parents/parentsStore.go:
```diff
@@ -145,8 +147,12 @@ func (p *parentsStore) UpdateParent(ctx context.Context, updateInfoFields []stri
 // If no parent ID is given, return all parents records in the parents table
 func (p parentsStore) ListParentsByID(ctx context.Context, parentIDs []int32) ([]Parent, error) {
+	if len(parentIDs) == 0 {
+		return nil, pdmerr.New("ParentStore: failed to list parents by ids - ids list is empty")
```
Let's return nil, nil error in this case as well.

sleepyzzzzzz replied:
changed. 
But have a question why should it not return an error? If there is no id in the parent_ids in the request param, Ruby returns all parents' records in the table. Then, does it mean using ListParentsByID in this case is not valid?

lhan-the-real commented:
I've referenced some of the code in this pr, and updated the implementation for ListAccountsByIDs here:
https://github.com/ParamountDataManagement/pdmgoauth/pull/111

Please take a look, and update the behavior based on the other change.

sleepyzzzzzz commented on internal/domain/parents/parentsStore.go:
```diff
@@ -160,25 +166,51 @@ func (p parentsStore) ListParentsByID(ctx context.Context, parentIDs []int32) ([
-		return nil, pdmerr.NewParent(err, "ParentStore: failed to listAllParents from DB - missing records")
+		invalidIDs := checkInValidParentID(parents, parentIDs)
+		invalidIDStr := strings.Trim(strings.Replace(fmt.Sprint(invalidIDs), " ", ", ", -1), "[]")
+		logrus.Warn(fmt.Sprintf("There exists parent ids does not in db: %s", invalidIDStr))
```
added

sleepyzzzzzz commented:
> I've referenced some of the code in this pr, and updated the implementation for ListAccountsByIDs here: #111
> 
> Please take a look, and update the behavior based on the other change.

One questions: 
In account store ListAccountsByID (line338-340), why len(listedAccounts) == 0 is not in the block for logging invalid ids warning?

lhan-the-real commented on apis/grpcservice/grpcusers/usersGRPCServer_test.go:
```diff
@@ -73,7 +73,6 @@ var _ = Describe("UsersGRPCServer", func() {
 			parentsStore,
 			mockHttpClient,
 		)
-		parentsStore = parents.CreateParentsStore(db)
```
Nice Catch!

lhan-the-real commented on internal/domain/parents/parentsStore.go:
```diff
@@ -145,40 +151,69 @@ func (p *parentsStore) UpdateParent(ctx context.Context, updateInfoFields []stri
+}
+
+func checkInValidParentID(resultParents []Parent, requestedIDs []int32) []int32 {
+	invalidIDs := []int32{}
```
var invalidIDs []int32

lhan-the-real commented on internal/domain/parents/parentsStore_test.go:
```diff
@@ -179,67 +187,74 @@ var _ = Describe("ParentsStore", func() {
 
 			Context("the parent IDs are partially valid", func() {
-				It("should throw error", func() {
+				It("should return the partially valid parent record and log the warning", func() {
```
If we are not able to verify log the warning, we should not put it into the description. Instead, we should put the expected return values there, which is: error to be nil.

lhan-the-real commented on internal/domain/parents/parentsStore_test.go:
```diff
@@ -179,67 +187,74 @@ var _ = Describe("ParentsStore", func() {
 
 			Context("page token is valid", func() {
-				It("should return all parents in the table limit by the page size and token", func() {
+				DescribeTable("return all parents in the table limit by the page size and token", func(
```
Can we also validate the listed parents struct values? This would help us to validate our query to be correct.

sleepyzzzzzz replied:
changed to expect equal created parents

lhan-the-real commented on internal/domain/parents/parentsStore_test.go:
```diff
@@ -19,7 +20,10 @@ var _ = Describe("ParentsStore", func() {
 
 	BeforeEach(func() {
-		parentsStore = parents.CreateParentsStore(db)
+		parentsStore = parents.CreateParentsStore(
```
Let's audit this test file, to ensure all queried results are fully validated for the outcome parent struct content.

sleepyzzzzzz replied:
changed to expect equal created parents

lhan-the-real commented:
> > I've referenced some of the code in this pr, and updated the implementation for ListAccountsByIDs here: #111
> > Please take a look, and update the behavior based on the other change.
> 
> One questions: In account store ListAccountsByID (line338-340), why len(listedAccounts) == 0 is not in the block for logging invalid ids warning?

When len == 0, it also fall into the bucket of `len(listedAccounts) != len(acocuntIDs)`. It should also log right?

sleepyzzzzzz commented:
> > > I've referenced some of the code in this pr, and updated the implementation for ListAccountsByIDs here: #111
> > > Please take a look, and update the behavior based on the other change.
> > 
> > 
> > One questions: In account store ListAccountsByID (line338-340), why len(listedAccounts) == 0 is not in the block for logging invalid ids warning?
> 
> When len == 0, it also fall into the bucket of `len(listedAccounts) != len(acocuntIDs)`. It should also log right?

Yes, my bad. Removed this check in parents now

lhan-the-real reviewed (approved):
Once the test is updated to compare the parentStruct, feel free to merge it in.
